File: client/src/core/scope.ts
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
// Copyright (c) 2026 2youg1 and the sprawling contributors

// The seam between the two spellings of one scope.
//
// A city has one notion of what a halt applies to and two ways of
// writing it down, by a decision recorded in `kernel::event::scope`:
// the Ledger holds `city`, `building:<addr>`, `workshop:<addr>`,
// because that is what every history already written holds, while a
// frame carries the tagged shape a sender would type. Both reach this
// client - answers in the frame shape, records in the Ledger shape -
// and this file is the one place they meet.
//
// Everything else in the client works in the frame shape, so nothing
// else compares strings. That is the defect this file exists to close:
// a page comparing a bare address against `building:<addr>` reported
// every shut building as running, and nothing failed.

import { Option, Schema } from "effect";

import { Address } from "../wire";
import type { HaltScope } from "../wire";

// The three words, as `kernel::event::scope` spells them, and the
// character that separates a kind from the address it names.
//
// `CITY` is exported because the two spellings agree on this one word:
// the Ledger writes `city` and a frame writes `city`, so a view that
// halts the whole city and a view that asks whether it is halted read
// one constant rather than each spelling it again.
export const CITY = "city";
const BUILDING = "building";
const WORKSHOP = "workshop";
const AT = ":";

// The address grammar, which `kernel::Address::parse` states for both
// ends. Decoding rather than asserting: the text after the separator is
// an address only if it is one, and a scope naming something else is a
// record this build cannot read.
const readAddress = Schema.decodeOption(Address);

// A scope as the Ledger spelled it, or null for a word this build does
// not know.
//
// Null rather than a default: a client that read an unknown scope as
// the city would report the whole city shut, and one that read it as
// nothing would show a stopped building as running. Neither is better
// than saying nothing.
export function scopeOf(spelled: string | null): HaltScope | null {
  if (spelled === null) return null;
  if (spelled === CITY) return CITY;
  const cut = spelled.indexOf(AT);
  if (cut < 0) return null;
  const kind = spelled.slice(0, cut);
  const addr = Option.getOrNull(readAddress(spelled.slice(cut + 1)));
  if (addr === null) return null;
  if (kind === BUILDING) return { building: addr };
  if (kind === WORKSHOP) return { workshop: addr };
  return null;
}

// Whether two scopes name the same thing.
//
// Structural, because a scope is a value: the tagged shape has no
// identity of its own, so two answers naming one building are two
// objects.
export function sameScope(left: HaltScope, right: HaltScope): boolean {
  if (left === CITY || right === CITY) return left === right;
  if ("building" in left && "building" in right) return left.building === right.building;
  if ("workshop" in left && "workshop" in right) return left.workshop === right.workshop;
  return false;
}

// Whether a list of shut scopes holds one building.
export function buildingIsShut(halted: readonly HaltScope[], addr: Address): boolean {
  return halted.some((held) => held !== CITY && "building" in held && held.building === addr);
}

// Whether a list of shut scopes holds the city itself.
export function cityIsShut(halted: readonly HaltScope[]): boolean {
  return halted.includes(CITY);
}

File: crates/kernel/src/event/scope.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
// Copyright (c) 2026 2youg1 and the sprawling contributors

//! What a standing decision applies to, as the ledger spells it.

use std::fmt;

use serde::{Deserialize, Serialize};

use crate::address::Address;
use crate::error::{AxCode, AxError};

/// The whole city, as a `city_halted` or `autonomy_changed` line spells
/// it.
const CITY: &str = "city";

/// The word before the address of a building.
const BUILDING: &str = "building";

/// The word before the address of a workshop.
const WORKSHOP: &str = "workshop";

/// What separates the kind of a scope from the address it names.
const AT: char = ':';

/// What a halt, a release or an autonomy change applies to.
///
/// **The wire form is what the histories already hold**: `city`,
/// `building:<addr>`, `workshop:<addr>`. That spelling used to be a
/// `format!` at one writer and a `split_once(':')` at each reader, so a
/// scope written one way and read another was one keystroke away. The
/// three words are here, the parse is here, and both ends of the ledger
/// call this type.
///
/// A word this build does not know is refused rather than read as some
/// other scope: a city that read an unknown scope as its own would
/// report itself shut, and one that read it as nothing would dispatch
/// into a scope a person stopped.
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Scope {
    /// Everything this city holds.
    City,
    /// One building and what is inside it.
    Building(Address),
    /// One workshop and what is inside it.
    Workshop(Address),
}

impl Scope {
    /// Whether a standing decision on this scope reaches an address.
    ///
    /// The city reaches everything; a building and a workshop reach
    /// what is within them, by the containment [`Address::is_within`]
    /// already decides, so "inside" means one thing in this city.
    #[must_use]
    pub fn covers(&self, addr: &Address) -> bool {
        match self {
            Scope::City => true,
            Scope::Building(scope) | Scope::Workshop(scope) => addr.is_within(scope),
        }
    }

    /// Reads back what [`fmt::Display`] wrote.
    ///
    /// # Errors
    /// `E_WIRE_MISMATCH` for a word that names none of the three
    /// shapes, and whatever [`Address::parse`] says about an address
    /// this city cannot hold.
    pub fn parse(raw: &str) -> Result<Scope, AxError> {
        if raw == CITY {
            return Ok(Scope::City);
        }
        let unknown = || {
            AxError::failure(AxCode::WireMismatch, "read which scope a line names", raw)
                .with_recovery(
                    "open this city with the build that wrote its history: this one knows \
                     `city`, `building:<address>` and `workshop:<address>`",
                )
        };
        let (kind, addr) = raw.split_once(AT).ok_or_else(unknown)?;
        let addr = Address::parse(addr)?;
        match kind {
            BUILDING => Ok(Scope::Building(addr)),
            WORKSHOP => Ok(Scope::Workshop(addr)),
            _ => Err(unknown()),
        }
    }
}

impl fmt::Display for Scope {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Scope::City => f.write_str(CITY),
            Scope::Building(addr) => write!(f, "{BUILDING}{AT}{addr}", addr = addr.as_str()),
            Scope::Workshop(addr) => write!(f, "{WORKSHOP}{AT}{addr}", addr = addr.as_str()),
        }
    }
}

impl Serialize for Scope {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        serializer.serialize_str(&self.to_string())
    }
}

impl<'de> Deserialize<'de> for Scope {
    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let raw = String::deserialize(deserializer)?;
        Scope::parse(&raw).map_err(serde::de::Error::custom)
    }
}

#[cfg(test)]
#[allow(
    clippy::unwrap_used,
    clippy::expect_used,
    clippy::panic,
    clippy::indexing_slicing,
    reason = "test code"
)]
mod tests {
    use super::*;

    #[test]
    fn the_three_scopes_write_the_words_the_histories_already_hold() {
        let building = Scope::Building(Address::parse("lab").unwrap());
        let workshop = Scope::Workshop(Address::parse("lab/bench").unwrap());
        assert_eq!(Scope::City.to_string(), "city");
        assert_eq!(building.to_string(), "building:lab");
        assert_eq!(workshop.to_string(), "workshop:lab/bench");
        for scope in [Scope::City, building, workshop] {
            assert_eq!(Scope::parse(&scope.to_string()).unwrap(), scope);
            assert_eq!(
                serde_json::to_string(&scope).unwrap(),
                format!("\"{scope}\"")
            );
        }
    }

    #[test]
    fn a_scope_reaches_what_is_inside_it_and_nothing_else() {
        let inside = Address::parse("lab/bench.md").unwrap();
        let elsewhere = Address::parse("hall/clerk.md").unwrap();
        let lab = Scope::Building(Address::parse("lab").unwrap());
        assert!(Scope::City.covers(&inside));
        assert!(lab.covers(&inside));
        assert!(!lab.covers(&elsewhere));
    }

    #[test]
    fn a_word_naming_no_scope_is_refused_rather_than_guessed() {
        for raw in ["lab", "borough:lab", "building:"] {
            assert!(Scope::parse(raw).is_err(), "{raw} should not read");
        }
    }
}
